File: remoter/src/main/java/remoter/compiler/builder/GenericParamBuilder.java
package remoter.compiler.builder;

import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterSpec;
import com.squareup.javapoet.TypeName;

import javax.annotation.processing.Messager;
import javax.lang.model.element.Element;
import javax.lang.model.element.VariableElement;
import javax.lang.model.type.TypeMirror;

/**
 * A {@link ParamBuilder} for generic type parameters
 */
class GenericParamBuilder extends ParamBuilder {


    protected GenericParamBuilder(Messager messager, Element element) {
        super(messager, element);
    }

    @Override
    public void writeParamsToProxy(ParameterSpec param, ParamType paramType, MethodSpec.Builder methodBuilder) {
        methodBuilder.addStatement("Class pClass$1L = getParcelerClass($1L)", param.name);
        methodBuilder.beginControlFlow("if (pClass$L != null)", param.name);
        methodBuilder.addStatement("data.writeInt(1)");
        methodBuilder.addStatement("data.writeString(pClass$L.getName())", param.name);
        methodBuilder.addStatement("org.parceler.Parcels.wrap(pClass$1L, $1L).writeToParcel(data, 0)", param.name);
        methodBuilder.endControlFlow();
        methodBuilder.beginControlFlow("else");
        methodBuilder.addStatement("data.writeInt(2)");
        methodBuilder.addStatement("data.writeValue($L)", param.name);
        methodBuilder.endControlFlow();
    }

    @Override
    public void readResultsFromStub(TypeName resultType, MethodSpec.Builder methodBuilder) {
        methodBuilder.addStatement("Class pClassResult = getParcelerClass(result)");
        methodBuilder.beginControlFlow("if (pClassResult != null)");
        methodBuilder.addStatement("reply.writeInt(1)");
        methodBuilder.addStatement("reply.writeString(pClassResult.getName())");
        methodBuilder.addStatement("org.parceler.Parcels.wrap(pClassResult, result).writeToParcel(reply, 0)");
        methodBuilder.endControlFlow();
        methodBuilder.beginControlFlow("else");
        methodBuilder.addStatement("reply.writeInt(2)");
        methodBuilder.addStatement("reply.writeValue(result)");
        methodBuilder.endControlFlow();
    }


    @Override
    public void readResultsFromProxy(TypeName resultType, MethodSpec.Builder methodBuilder) {
        methodBuilder.beginControlFlow("if (reply.readInt() == 1)");
        methodBuilder.addStatement("result = ($T)getParcelerObject(reply.readString(), reply)", resultType);
        methodBuilder.endControlFlow();
        methodBuilder.beginControlFlow("else");
        methodBuilder.addStatement("result = (" + resultType.toString() + ")reply.readValue(getClass().getClassLoader())");
        methodBuilder.endControlFlow();
    }

    @Override
    public void writeParamsToStub(ParameterSpec param, ParamType pType, MethodSpec.Builder methodBuilder) {
        super.writeParamsToStub(param, pType, methodBuilder);
        methodBuilder.beginControlFlow("if (data.readInt() == 1)");
        methodBuilder.addStatement("$L = ($T)getParcelerObject(data.readString(), data)", param.name, param.type);
        methodBuilder.endControlFlow();
        methodBuilder.beginControlFlow("else");
        methodBuilder.addStatement("$L = ($T)data.readValue(getClass().getClassLoader())", param.name, param.type);
        methodBuilder.endControlFlow();
    }
}

File: remoter/src/main/java/remoter/compiler/kbuilder/GenericParamBuilder.kt
package remoter.compiler.kbuilder

import com.squareup.kotlinpoet.FunSpec
import javax.lang.model.element.Element
import javax.lang.model.element.ExecutableElement
import javax.lang.model.element.VariableElement
import javax.lang.model.type.TypeMirror


/**
 * A [ParamBuilder] for generic type parameters
 */
internal class GenericParamBuilder(remoterInterfaceElement: Element, bindingManager: KBindingManager) : ParamBuilder(remoterInterfaceElement, bindingManager) {
    override fun writeParamsToProxy(param: VariableElement, paramType: ParamType, methodBuilder: FunSpec.Builder) {
        methodBuilder.addStatement("val pClass" + param.simpleName + " = getParcelerClass(" + param.simpleName + ")")
        methodBuilder.beginControlFlow("if (pClass" + param.simpleName + " != null)")
        methodBuilder.addStatement("$DATA.writeInt(1)")
        methodBuilder.addStatement("$DATA.writeString(pClass" + param.simpleName + ".getName())")
        methodBuilder.addStatement("org.parceler.Parcels.wrap(pClass" + param.simpleName + ", " + param.simpleName + ").writeToParcel($DATA, 0)")
        methodBuilder.endControlFlow()
        methodBuilder.beginControlFlow("else")
        methodBuilder.addStatement("$DATA.writeInt(2)")
        methodBuilder.addStatement("$DATA.writeValue(" + param.simpleName + " )")
        methodBuilder.endControlFlow()
    }

    override fun readResultsFromStub(methodElement: ExecutableElement, resultType: TypeMirror, methodBuilder: FunSpec.Builder) {
        methodBuilder.addStatement("val pClassResult = getParcelerClass($RESULT)")
        methodBuilder.beginControlFlow("if (pClassResult != null)")
        methodBuilder.addStatement("$REPLY.writeInt(1)")
        methodBuilder.addStatement("$REPLY.writeString(pClassResult.getName())")
        methodBuilder.addStatement("org.parceler.Parcels.wrap(pClassResult, $RESULT).writeToParcel($REPLY, 0)")
        methodBuilder.endControlFlow()
        methodBuilder.beginControlFlow("else")
        methodBuilder.addStatement("$REPLY.writeInt(2)")
        methodBuilder.addStatement("$REPLY.writeValue($RESULT)")
        methodBuilder.endControlFlow()
    }

    override fun readResultsFromProxy(methodType: ExecutableElement, methodBuilder: FunSpec.Builder) {
        val resultType = methodType.getReturnAsKotlinType()
        methodBuilder.beginControlFlow("if ($REPLY.readInt() == 1)")
        methodBuilder.addStatement("$RESULT = getParcelerObject($REPLY.readString(), $REPLY) as %T", resultType)
        methodBuilder.endControlFlow()
        methodBuilder.beginControlFlow("else")
        methodBuilder.addStatement("$RESULT = $REPLY.readValue(javaClass.getClassLoader()) as %T", resultType)
        methodBuilder.endControlFlow()
    }

    override fun writeParamsToStub(methodType: ExecutableElement, param: VariableElement, paramType: ParamType, paramName: String, methodBuilder: FunSpec.Builder) {
        super.writeParamsToStub(methodType, param, paramType, paramName, methodBuilder)
        val paramTypeString = param.asType().toString()
        methodBuilder.beginControlFlow("if ($DATA.readInt() == 1)")
        methodBuilder.addStatement("$paramName = getParcelerObject($DATA.readString(), $DATA) as %T", param.asKotlinType())
        methodBuilder.endControlFlow()
        methodBuilder.beginControlFlow("else")
        methodBuilder.addStatement("$paramName = $DATA.readValue(javaClass.getClassLoader()) as %T", param.asKotlinType())
        methodBuilder.endControlFlow()
    }
}


